File: app/csharp/simple/Services/AiSearchService.cs
using System;
using System.Collections.Generic;

using Azure;
using Azure.Search.Documents;
using Azure.Search.Documents.Indexes;
using Azure.Search.Documents.Models;
using Simple.Services.Helpers;

namespace Simple.Services
{
    /// <summary>
    /// Azure AI Searchに接続するためのSearchClientを構築します。
    /// </summary>
    public class SearchClientBuilder
    {
        /// <summary>
        /// SearchClientインスタンスを構築して返します。
        /// </summary>
        /// <returns>SearchClientインスタンス。</returns>
        /// <exception cref="InvalidOperationException">検索サービスの設定が不足している場合にスローされます。</exception>
        public SearchClient BuildSearchClient(string indexName)
        {
            var searchServiceName = Environment.GetEnvironmentVariable("AI_SEARCH_SERVICE_NAME");
            var apiKey = Environment.GetEnvironmentVariable("AI_SEARCH_API_KEY");

            if (string.IsNullOrEmpty(searchServiceName) || string.IsNullOrEmpty(indexName) || string.IsNullOrEmpty(apiKey))
            {
                throw new InvalidOperationException("Search service configuration is missing.");
            }

            var serviceEndpoint = $"https://{searchServiceName}.search.windows.net/";
            var credential = new AzureKeyCredential(apiKey);

            return new SearchClient(new Uri(serviceEndpoint), indexName, credential);
        }
    }

    /// <summary>
    /// Azure AI Searchを使用してAI検索サービスを提供します。
    /// </summary>
    public class AiSearchService
    {
        private readonly SearchClientBuilder _searchClientBuilder;

        /// <summary>
        /// <see cref="AiSearchService"/> クラスの新しいインスタンスを初期化します。
        /// </summary>
        /// <param name="searchClientBuilder">SearchClientBuilderインスタンス。</param>
        public AiSearchService(SearchClientBuilder searchClientBuilder)
        {
            _searchClientBuilder = searchClientBuilder;
        }

        /// <summary>
        /// 指定されたクエリを使用して全文検索を実行します。
        /// </summary>
        /// <param name="query">検索クエリ。</param>
        /// <returns>検索ドキュメントのリスト。</returns>
        public List<SearchDocument> FullTextSearch(string query)
        {
            var searchClient = _searchClientBuilder.BuildSearchClient(Environment.GetEnvironmentVariable("AI_SEARCH_INDEX_NAME"));
            var options = new SearchOptions
            {
                Size = 5
            };
            options.Select.Add("keyphrases");
            options.Select.Add("content");
            options.Select.Add("text");
            SearchResults<SearchDocument> results = searchClient.Search<SearchDocument>(query, options);

            List<SearchDocument> searchResults = new List<SearchDocument>();
            foreach (SearchResult<SearchDocument> result in results.GetResults())
            {
                searchResults.Add(result.Document);
            }
            return searchResults;
        }

        public List<SearchDocument> VectorSearch(ReadOnlyMemory<float> vectorizedQuery)
        {
            var searchClient = _searchClientBuilder.BuildSearchClient(Environment.GetEnvironmentVariable("AI_SEARCH_VECTOR_INDEX_NAME"));
            var options = new SearchOptions
            {
                Size = 5,
                VectorSearch = new()
                {
                    Queries = {
                        new VectorizedQuery(vectorizedQuery)
                        {
                            KNearestNeighborsCount = 3,
                            Fields = { "text_vector" }
                        }
                    }
                }
            };
            SearchResults<SearchDocument> results = searchClient.Search<SearchDocument>(options);

            List<SearchDocument> searchResults = new List<SearchDocument>();
            foreach (SearchResult<SearchDocument> result in results.GetResults())
            {
                searchResults.Add(result.Document);
                Console.WriteLine(result.Document); 
            }
            return searchResults;
        }
    }
}

File: app/python/simple/service/ai_search_service.py
import os
from typing import List

from azure.core.credentials import AzureKeyCredential

from azure.search.documents import SearchClient
from azure.search.documents.models import VectorizedQuery

from .helpers.helper_methods import get_secret_from_key_vault


def build_search_client():
    """
    SearchClientインスタンスを構築して返します。

    この関数は、環境変数から検索サービス名、インデックス名、およびAPIキーを取得し、
    サービスエンドポイントを構築して、指定されたサービスエンドポイント、
    インデックス名、およびAPIキーで構成されたSearchClientインスタンスを返します。

    Returns:
        SearchClient: 指定されたサービスエンドポイント、インデックス名、
        およびAPIキーで構成されたSearchClientのインスタンス。
    """
    search_service_name = os.environ.get("AI_SEARCH_SERVICE_NAME")
    index_name = os.environ.get("AI_SEARCH_INDEX_NAME")
    api_key = os.environ.get("AI_SEARCH_API_KEY")
    service_endpoint = "https://{0}.search.windows.net/".format(search_service_name)
    return SearchClient(service_endpoint, index_name, AzureKeyCredential(api_key))


def search_fulltext(query: str) -> list:
    """
    フルテキスト検索を実行して結果を返します。

    この関数は、指定されたクエリを使用してAzure AI Searchサービスに
    リクエストを送信し、検索結果を返します。

    Args:
        query (str): 検索クエリ。

    Returns:
        list: 検索結果のリスト。
    """
    search_client = build_search_client()
    results = search_client.search(search_text=query, select=["keyphrases", "content", "text"], top=5)
    return [result for result in results]

def search_vector(vector: List[float]) -> list:
    """
    ベクトル検索を実行して結果を返します。

    この関数は、指定されたクエリを使用してAzure AI Searchサービスに
    リクエストを送信し、ベクトル検索の結果を返します。

    Args:
        vector (List[float]): ベクトル化された検索クエリ。

    Returns:
        list: 検索結果のリスト。
    """
    search_client = build_search_client()
    results = search_client.search(
        search_text=None, 
        vector_queries=[
            VectorizedQuery(
                kind="vector", vector=vector, k_nearest_neighbors=3
            )
        ],
    )
    return [result for result in results]